server/lib/chat/chat-api.ts

```diff
@@ -0,0 +1,206 @@
+    case ChatServiceErrorCode.UserOffline:
+      throw new httpErrors.NotFound(err.message)
+    case ChatServiceErrorCode.InvalidJoinAction:
+    case ChatServiceErrorCode.InvalidLeaveAction:
+    case ChatServiceErrorCode.InvalidSendAction:
+    case ChatServiceErrorCode.InvalidGetHistoryAction:
+    case ChatServiceErrorCode.InvalidGetUsersAction:
+      throw new httpErrors.BadRequest(err.message)
+    case ChatServiceErrorCode.LeaveShieldBattery:
+      throw new httpErrors.Forbidden(err.message)
+    default:
+      assertUnreachable(err.code)
+  }
+}
+
+@httpApi()
+export class ChatApi extends HttpApi {
+  constructor() {
+    super('/chat')
+    container.resolve(ChatService)
+  }
+
+  applyRoutes(router: Router): void {
+    router
+      .use(featureEnabled(MULTI_CHANNEL), ensureLoggedIn)
+      .post(
+        '/:channelName',
+        throttleMiddleware(joinThrottle, ctx => String(ctx.session!.userId)),
+        joinChannel,
+      )
+      .delete(
+        '/:channelName',
+        throttleMiddleware(leaveThrottle, ctx => String(ctx.session!.userId)),
+        leaveChannel,
+      )
+      .patch(
+        '/:channelName',
+        throttleMiddleware(sendThrottle, ctx => String(ctx.session!.userId)),
+        sendChatMessage,
+      )
+      .get(
+        '/:channelName/chat',
+        throttleMiddleware(retrievalThrottle, ctx => String(ctx.session!.userId)),
+        getChannelHistory,
+      )
+      .get(
+        '/:channelName/users',
+        throttleMiddleware(retrievalThrottle, ctx => String(ctx.session!.userId)),
+        getChannelUsers,
+      )
+  }
+}
+
+function joinChannel(ctx: RouterContext) {
+  const {
+    params: { channelName },
+  } = validateRequest(ctx, {
+    params: Joi.object<{ channelName: string }>({
+      channelName: Joi.string().max(CHANNEL_MAXLENGTH).pattern(CHANNEL_PATTERN).required(),
+    }),
+  })
+
+  try {
+    const chatService = container.resolve(ChatService)
+    chatService.joinChannel(channelName, ctx.session!.userName)
+
+    ctx.status = 204
+  } catch (err) {
+    convertChatServiceError(err)
+  }
+}
+
+function leaveChannel(ctx: RouterContext) {
+  const {
+    params: { channelName },
+  } = validateRequest(ctx, {
+    params: Joi.object<{ channelName: string }>({
+      channelName: Joi.string().max(CHANNEL_MAXLENGTH).pattern(CHANNEL_PATTERN).required(),
+    }),
+  })
+
+  try {
+    const chatService = container.resolve(ChatService)
+    chatService.leaveChannel(channelName, ctx.session!.userName)
+
+    ctx.status = 204
+  } catch (err) {
+    convertChatServiceError(err)
+  }
+}
+
+function sendChatMessage(ctx: RouterContext) {
+  const {
+    params: { channelName },
+    body: { message },
+  } = validateRequest(ctx, {
+    params: Joi.object<{ channelName: string }>({
+      channelName: Joi.string().max(CHANNEL_MAXLENGTH).pattern(CHANNEL_PATTERN).required(),
+    }),
+    body: Joi.object<{ message: string }>({
+      message: Joi.string().min(1).required(),
```

Comment:
Pretty sure this has a max length too?

Reply:
Yeah, but we're simply cutting it off here: https://github.com/ShieldBattery/ShieldBattery/blob/master/server/lib/messaging/filter-chat-message.js instead of throwing an error.